src/transform/src/predicate_propagation.rs

```diff
@@ -0,0 +1,381 @@
+                Vec::new()
+            }
+            RelationExpr::Let { id, value, body } => {
+                // This deals with shadowed let bindings, but perhaps we should just complain instead?
+                let value_knowledge = PredicateKnowledge::action(value, let_knowledge)?;
+                let prior_knowledge =
+                    let_knowledge.insert(expr::Id::Local(id.clone()), value_knowledge);
+                let body_knowledge = PredicateKnowledge::action(body, let_knowledge)?;
+                let_knowledge.remove(&expr::Id::Local(id.clone()));
+                if let Some(prior_knowledge) = prior_knowledge {
+                    let_knowledge.insert(expr::Id::Local(id.clone()), prior_knowledge);
+                }
+                body_knowledge
+            }
+            RelationExpr::Project { input, outputs } => {
+                let mut input_knowledge = PredicateKnowledge::action(input, let_knowledge)?;
+                // Need to
+                // 1. restrict to supported columns,
+                // 2. rewrite constraints using new column identifiers,
+                // 3. add equivalence constraints for repeated columns.
+                let mut remap = HashMap::new();
+                for (index, column) in outputs.iter().enumerate() {
+                    if !remap.contains_key(column) {
+                        remap.insert(column, index);
+                    }
+                }
+
+                // 1. restrict to supported columns,
+                input_knowledge
+                    .retain(|expr| expr.support().into_iter().all(|c| remap.contains_key(&c)));
+
+                // 2. rewrite constraints using new column identifiers,
+                for predicate in input_knowledge.iter_mut() {
+                    predicate.visit_mut(&mut |e| {
+                        if let ScalarExpr::Column(c) = e {
+                            *c = remap[c];
+                        }
+                    })
+                }
+
+                // 3. add equivalence constraints for repeated columns.
+                for (index, column) in outputs.iter().enumerate() {
+                    if remap[column] != index {
+                        input_knowledge.push(
+                            ScalarExpr::Column(remap[column])
+                                .call_binary(ScalarExpr::Column(*column), expr::BinaryFunc::Eq),
+                        );
+                    }
+                }
+                input_knowledge
+            }
+            RelationExpr::Map { input, scalars } => {
+                let input_knowledge = PredicateKnowledge::action(input, let_knowledge)?;
+                // Scalars could be simplified based on known predicates.
+                for scalar in scalars.iter_mut() {
+                    optimize(scalar, &input.typ(), &input_knowledge[..])?;
+                }
+                input_knowledge
```

Comment:
would it be helpful to add `ScalarExpr::Column(arity) = scalars[0]`, `ScalarExpr::Column(arity+1)=scalars[1]`, etc. into `input_knowledge`? Or do we believe other transforms successfully incorporate this knowledge already?

Reply:
Yes. This is probably the right way to recover `map_lifting.rs` behavior.